### DISCUSSION TITLE
BiDi support

### DISCUSSION THREAD

**Discussion by Piedone** [atata-framework/atata#847] [General]
BiDi is now available in Selenium, see e.g. [here](https://github.com/SeleniumHQ/selenium/pull/14318/files#diff-74a59237dd909eaf45f31d8157e01415e5ddd6cd33066dd22a753f7741175769). This allows things like:

```c#
var context = await driver.AsBiDiContextAsync();
await context.NavigateAsync("url");
```

Will you adopt Atata somehow, @YevgeniyShunevych? Do we have an advantage over doing something like `driver.Navigate().GoToUrl("url");`?

Thank you!

**Comment by YevgeniyShunevych** [comment]
Right @Piedone, BiDi is a great and significant evolution step of WebDriver and it continues to cover old functionality and bring new.

Regarding navigation. I performed some test locally for Chrome and Firefox with headed and headless modes. At the moment I don't see a performance difference. Also I need to mention an important thing which might be unobvious, `driver.Navigate().GoToUrl("url")` isn't equivalent to `await context.NavigateAsync("url")`, but equivalent to `await context.NavigateAsync(url, new NavigateOptions { Wait = ReadinessState.Complete })`. By default, when `NavigateOptions.Wait` is not specified `ReadinessState.None` is used, which is equivalent to `DriverOptions.PageLoadStrategy = PageLoadStrategy.None`. [`PageLoadStrategy`](https://www.selenium.dev/documentation/webdriver/drivers/options/#pageloadstrategy) on the other side is set to `Normal` by default, and `GoToUrl` uses that setting.

By the way you can try to experiment with `PageLoadStrategy` option, for example setting `Eager` value. It may improve overall tests speed a bit, but be careful to not increase tests flakiness.

Regarding Atata and BiDi adoption. I want and will need to migrate Atata internal functionality to BiDi. One day, as far as I understand, BiDi is planned to replace classic WebDriver API. But today BiDi is not feature complete yet, still in development. Also I'm currently busy with Atata 4, which already brings a lot of changes (including breaking ones). So for Atata 4 I don't want to bring any WebDriver-related changes, because migration to newer version can become even harder. But you are absolutely free to use BiDi features for your needs even with the current version of Atata.

**Reply by Piedone** [reply]
Thanks for elaborating it in such detail!

Yeah, supposedly, eventually all the current WebDriver methods will use BiDi under the hood too.

One thing I'm curious and concerned about is async: all the BiDi APIs are async, so (unless anybody wants to do sync-over-async) the WebDriver API should also be async, and its consumers too. In the Lombiq UI Testing Toolbox almost everything is async anyway, but still, this can have a rippling effect. It's also something that can help the throughput of test execution.
